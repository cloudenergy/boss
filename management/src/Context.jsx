import React from 'react'
import { Subject } from 'rxjs-compat'

export const ActionContext = React.createContext(new Subject());
