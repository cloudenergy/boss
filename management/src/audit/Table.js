import {AuditAction} from '../Action'
import React from 'react'
import {AuditContext} from '../Context'
import $ from 'jquery'
const {Context} = AuditContext
const Table = ({data, title}) => (
  <Context.Consumer>{({table, actions, modalId, idLens, statusLens, color})=>(
    <div className="accordion" id="banking-audit">
      <div className="card">
        <div className={"card-header " + color}>
          <h5 className="mb-0">
            {title}
          </h5>
        </div>
        <div>
          <div className="card-body">
            <div className="form-group">
              <input className="form-control col-2" type="search" placeholder="搜索" aria-label="Search" onChange={e=> actions.next(AuditAction.Query(e.target.value))} />
            </div>
            <table className="table table-hover">
              <thead className="sticky-top">
                <tr>
                  {
                    table.map((col, i) =>(
                      <th key={i} scope="col">{col.name}</th>
                    ))
                  }
                </tr>
              </thead>
              <tbody>
                {
                  data.map((project,index)=>(
                    <tr data-toggle="modal" key={index} onClick={()=>{
                        actions.next(AuditAction.Popup(idLens(project), statusLens(project)))
                    }}>
                      {table.map((col,index)=>(
                        <td key={index}>{col.lens(project)}</td>
                      ))}
                    </tr>
                  ))
                }
              </tbody>
            </table>
          </div>
        </div>
      </div>
    </div>
  )}
  </Context.Consumer>
)
export default Table
