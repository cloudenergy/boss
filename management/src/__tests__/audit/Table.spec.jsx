import React from 'react';
import Table from '../../audit/Table';
import renderer from 'react-test-renderer';
import * as r from 'ramda'
import Enzyme, {mount} from 'enzyme'
import Adapter from 'enzyme-adapter-react-16'
import {AuditContext,createActionContext} from '../../Context'
const {Context} = AuditContext

Enzyme.configure({ adapter: new Adapter() })

describe('Table', ()=>{
  const tableLayout = [{
    name: 'property',
    lens: r.view(r.lensPath(['a', 'b']))
  },{
    name: 'property2',
    lens: r.view(r.lensPath(['c']))
  }]
  const data = [{
    a:{b:1.1},
    c:1.2
  },{
    a:{b:2.1},
    c:2.2
  }]
  let subject, Var;
  beforeEach(()=>{
    Var = createActionContext().Var
    subject = (<Context.Provider value={{actions: Var,
                                         table:tableLayout,
                                         modalId: "modalId",
                                         idLens: ()=>999,
                                         statusLens: ()=>false
    }}>
      <Table data={data} />
    </Context.Provider>)
  })

  it('renders a Table base on context', () => {
    expect(renderer.create(subject).toJSON()).toMatchSnapshot()
  })

  it('trigger popup', (done,fail) => {
    let wrapper = mount(subject)
    Var.subscribe(actions => actions.case({
      Popup: (id,status) => {
        expect(id).toBe(999)
        expect(status).toBe(false)
        wrapper.unmount()
        done()
      },
      _: done.fail
    }))
    wrapper.find('tbody tr').first().simulate('click')
  })

  it('trigger search', (done) => {
    let wrapper = mount(subject)
    Var.subscribe(actions => actions.case({
      Query: (str) => {
        expect(str).toBe('1.2')
        done()
      },
      _: done.fail
    }))
    wrapper.find('input[type="search"]').first().simulate('change', {target:{value:'1.2'}})
  })

})
