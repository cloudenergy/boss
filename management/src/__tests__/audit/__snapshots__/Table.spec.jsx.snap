// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Table renders a Table base on context 1`] = `
<div
  className="accordion"
  id="banking-audit"
>
  <div
    className="card"
  >
    <div
      className="card-header undefined"
    >
      <h5
        className="mb-0"
      />
    </div>
    <div>
      <div
        className="card-body"
      >
        <div
          className="form-group"
        >
          <input
            aria-label="Search"
            className="form-control col-2"
            onChange={[Function]}
            placeholder="搜索"
            type="search"
          />
        </div>
        <table
          className="table table-hover"
        >
          <thead
            className="sticky-top"
          >
            <tr>
              <th
                scope="col"
              >
                property
              </th>
              <th
                scope="col"
              >
                property2
              </th>
            </tr>
          </thead>
          <tbody>
            <tr
              data-target="#modalId"
              data-toggle="modal"
              onClick={[Function]}
            >
              <td>
                1.1
              </td>
              <td>
                1.2
              </td>
            </tr>
            <tr
              data-target="#modalId"
              data-toggle="modal"
              onClick={[Function]}
            >
              <td>
                2.1
              </td>
              <td>
                2.2
              </td>
            </tr>
          </tbody>
        </table>
      </div>
    </div>
  </div>
</div>
`;
